### PR TITLE
License filter and detailed output from license service

### PR DESCRIPTION
**Pull request by harjinder-hari**
Hi @miteshvp @pkajaba ,

Can you please review the license related changes in the worker ? There are mainly two changes:

1. Stack Aggregator v2 ( stackaggregator_v2.py ) now extracts details like unknown licenses, conflicting licenses, license outlier packages, etc. from the license analysis REST service.

2. Recommender ( recommender.py ) now filters PGM recommendations based on license i.e. we avoid recommending any package(s) that might result in conflict later if user chooses that package.

Thanks,
Harjinder

### CODE REVIEW

**Comment by miteshvp** [comment]
[test]

**Comment by jpopelka** [comment]
@harjinder-hari we've recently enabled the linter in CI and your code has [some minor issues](https://ci.centos.org/job/devtools-fabric8-analytics-worker-fabric8-analytics-pylint/31/console)

File: f8a_worker/workers/recommender.py [hunk 1]
```diff
@@ -639,6 +642,78 @@ def _get_stack_values(self, similar_stacks_list):
...
+        "companion_packages": companion_packages
+    }
+
+    resp = {}
+    try:
+        license_req = get_session_retry().post(license_url, data=json.dumps(payload))
+        resp = license_req.json()
+    except:  # TODO customized exception
+        msg = traceback.format_exc()
+        _logger.error("Unexpected error happened while invoking license analysis!\n{}".format(msg))
```
Comment (jpopelka) [inline f8a_worker/workers/recommender.py]: I think `_logger.exception("Unexpected error happened while invoking license analysis!")` should do the same, see https://stackoverflow.com/questions/5191830/best-way-to-log-a-python-exception

File: f8a_worker/workers/recommender.py [hunk 2]
```diff
@@ -639,6 +642,78 @@ def _get_stack_values(self, similar_stacks_list):
...
+        license_filter = la_output.get('license_filter', {})
+        conflict_packages_alt = license_filter.get('alternate_packages', {}).get('conflict_packages', [])
+        conflict_packages_com = license_filter.get('companion_packages', {}).get('conflict_packages', [])
+
+    list_pkg_names_alt = []
+    for epv in epv_list_alt:
+        name = epv.get('pkg', {}).get('name', [''])[0]
+        if name in conflict_packages_alt:
+            list_pkg_names_alt.append(name)
+            epv_list_alt.remove(epv)
```
Comment (jpopelka) [inline f8a_worker/workers/recommender.py]: I'm not sure it's a good idea to remove items from list while iterating over it.
In your case I'd probably use `for epv in epv_list_alt[:]:` (so you iterate over a copy of `epv_list_alt`)

See https://stackoverflow.com/questions/1207406/remove-items-from-a-list-while-iterating

File: f8a_worker/workers/recommender.py [hunk 3]
```diff
@@ -639,6 +642,78 @@ def _get_stack_values(self, similar_stacks_list):
...
+
+    list_pkg_names_com = []
+    for epv in epv_list_com:
+        name = epv.get('pkg', {}).get('name', [''])[0]
+        if name in conflict_packages_com:
+            list_pkg_names_com.append(name)
+            epv_list_com.remove(epv)
```
Comment (jpopelka) [inline f8a_worker/workers/recommender.py]: The same as above.

**Comment by jpopelka** [comment]
[test]

**Comment by jpopelka** [comment]
https://ci.centos.org/job/devtools-fabric8-analytics-worker-fabric8-analytics-pylint/38/console doesn't look much better. Please run `run-linter.sh` before you push next time to check that all the warnings are fixed.

**Comment by harjinder-hari** [comment]
Yes @jpopelka I had run the run-linter.sh before pushing the changes. It is really strange to see those linter errors !!

For example, the line#13 is not 104 characters long but still got this error:
f8a_worker/workers/recommender.py:13:101: E501 line too long (104 > 100 characters)

When I run the linter in my computer, I got "All checks passed" message !

Let me dig into details.

File: f8a_worker/workers/recommender.py (hunk 1 above)
Comment (harjinder-hari) [reply f8a_worker/workers/recommender.py]: Done. I tested locally and confirmed that `_logger.exception()` works as expected.

File: f8a_worker/workers/recommender.py (hunk 2 above)
Comment (harjinder-hari) [reply f8a_worker/workers/recommender.py]: Done. I now use for `epv in epv_list_alt[:]:`.

File: f8a_worker/workers/recommender.py (hunk 3 above)
Comment (harjinder-hari) [reply f8a_worker/workers/recommender.py]: Done. I now use `for epv in epv_list_com[:]:`

**Comment by harjinder-hari** [comment]
[test]

**Comment by harjinder-hari** [comment]
It seems that CI-CD linters are considering different code ... not sure why !?!
Actually, the same linter works fine in my computer. I will look into this tomorrow.

**Comment by jpopelka** [comment]
[test]

**Comment by jpopelka** [comment]
> When I run the linter in my computer, I got "All checks passed" message !

Me too, I'll check what causes the difference.

**Comment by harjinder-hari** [comment]
Thanks a lot @jpopelka ! That will help as I have no clue why linter in commit-build is considering different src code.

**Comment by jpopelka** [comment]
I don't understand but, but even github claims there are **no** merge conflicts, there actually seem to be some

see [PR build](https://ci.centos.org/job/devtools-fabric8-analytics-worker-fabric8-analytics/157/console)
```
E     File "/f8a_worker/f8a_worker/workers/recommender.py", line 12
E       <<<<<<< HEAD
E        ^
E   SyntaxError: invalid syntax
```

or [pylint run](https://ci.centos.org/job/devtools-fabric8-analytics-worker-fabric8-analytics-pylint/40/console)
```
+ git rebase --preserve-merges origin/master
Rebasing (1/3)
Removing tests/workers/test_sentiment_scorer.py
Removing tests/test_conf.py
Removing hack/env.sh
Auto-merging f8a_worker/workers/stackaggregator_v2.py
CONFLICT (content): Merge conflict in f8a_worker/workers/stackaggregator_v2.py
Removing f8a_worker/workers/sentiment_scorer.py
Auto-merging f8a_worker/workers/recommender.py
CONFLICT (content): Merge conflict in f8a_worker/workers/recommender.py
Removing f8a_worker/defaultconf.py
Removing f8a_worker/conf.py
Automatic merge failed; fix conflicts and then commit the result.
Error redoing merge eb2c26
```

**Comment by jpopelka** [comment]
Can you do `git rebase master` to see if that helps ?

**Comment by harjinder-hari** [comment]
@jpopelka ,
I am trying the rebase option. But it is giving me too many conflicts to resolve.
Can I create a fresh branch on top of latest commit from upstream/master and create a fresh PR ?

**Comment by jpopelka** [comment]
Yes, you can create new PR

**Comment by harjinder-hari** [comment]
Hey @jpopelka ... sorry for the confusion ... but I have rebased and pushed again. Let us hope it helps the linters.

**Comment by harjinder-hari** [comment]
[test]

**Comment by jpopelka** [comment]
[test]

**Pull request merged**